File: kraken/src/main/java/glitch/kraken/services/ClipService.java
package glitch.kraken.services;

import glitch.api.http.Routes;
import glitch.auth.objects.json.Credential;
import glitch.kraken.GlitchKraken;
import glitch.kraken.object.json.Clip;
import glitch.kraken.services.request.FollowedClipsRequest;
import glitch.kraken.services.request.TopClipsRequest;
import glitch.service.AbstractHttpService;
import reactor.core.publisher.Mono;

public class ClipService extends AbstractHttpService {
    public ClipService(GlitchKraken rest) {
        super(rest.getClient(), rest.getHttpClient());
    }

    public Mono<Clip> getClip(String slug) {
        return exchangeTo(Routes.get("/clips/%s").newRequest(slug), Clip.class);
    }

    public TopClipsRequest getTopClips() {
        return new TopClipsRequest(http);
    }

    public FollowedClipsRequest getFollowedClips(Credential credential) {
        return new FollowedClipsRequest(http, credential);

    }
}

File: kraken/src/main/kotlin/io/glitchlib/v5/service/ClipService.kt
package io.glitchlib.v5.service

import io.glitchlib.GlitchClient
import io.glitchlib.auth.Credential
import io.glitchlib.auth.Scope
import io.glitchlib.internal.http.bodySingle
import io.glitchlib.model.CursorList
import io.glitchlib.v5.internal.AbstractKrakenService
import io.glitchlib.v5.model.json.Clip
import io.glitchlib.v5.service.request.FollowedClipsRequest
import io.glitchlib.v5.service.request.TopClipsRequest

class ClipService(client: GlitchClient) : AbstractKrakenService(client) {

    fun getTopClips(request: TopClipsRequest.() -> Unit = {}) =
        get<CursorList<Clip>>("/clips/top", TopClipsRequest().apply(request)()).bodySingle

    fun getClip(slug: String) =
        get<Clip>("/clips/$slug").bodySingle

    fun getFollowedClips(credential: Credential, request: FollowedClipsRequest.() -> Unit = {}) =
        if (credential.scopeCheck(Scope.USER_READ))
            get<CursorList<Clip>>("/clips/followed", FollowedClipsRequest(credential).apply(request)()).bodySingle
        else scopeIsMissing(Scope.USER_READ)
}
